[MRG+1] Implements SelectDimensionKernel

MechCoder opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

Finishes up and closes https://github.com/scikit-learn/scikit-learn/pull/7070

Fixes #6868

#### What does this implement/fix? Explain your changes.

I just made some minor cosmetic changes to an example.

#### Any other comments?

The original PR already had two +1's by me and by @jmetzen . This should be good to merge once tests pass.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

btabibian committed:
Implements SelectDimensionKernel

MechCoder committed:
Cosmetics

MechCoder commented:
I wanted to merge as soon as tests pass but running the example gives me this warning.

```
Estimator SelectDimensionKernel modifies parameters in __init__. This behavior is deprecated as of 0.18 and support for this behavior will be removed in 0.20.
```

Will investigate in a while.

btabibian commented:
@MechCoder It seems it is due to this: 
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/base.py#L110

Which I don't quite understand. if `equality_test` is True, it produces a warning. If it is False it produces an Error.

MechCoder committed:
Remove pointing to kernel through kernel_ and all references to kernel_

MechCoder commented:
@btabibian I believe that the test is correct. But the parameters are not modified at init time but at `get_params` time. See

```python
rbf = RBF()
sdk = SelectDimensionKernel(rbf, active_dims=[1])
print(sdk.get_params()["kernel"] is rbf)
```

Creating a `kernel_` attribute to track the hyperparameters of `kernel`, creates issues such as these:

```python
sdk.set_params(kernel=ExpSineSquared())
SelectDimensionKernel(kernel: RBF(length_scale=1), active dimensions:[1])
sdk.kernel
ExpSineSquared(length_scale=1, periodicity=1)
sdk.kernel_
RBF(length_scale=1)
```

I removed the references to `kernel_` and modify the hyperparameters of `kernel` directly at `set_params`, and `theta.setter` which fixes the issue. Perhaps you could remind me why that you chose to have `kernel_` in your initial pull request instead of modifying the hyperparameters of `kernel`?

MechCoder committed:
Implements the following:

1. Fix a bug when 0 is present as an active index.
2. Better error message when active_dims size != length_scale
3. Add more tests

MechCoder commented:
I further fixed another bug is 0 is provided as an active_dim due to this line.

```python
active_dims = np.where(active_dims)[0]
```

added a better error message when the length_scale size is not the same as the size of the active dimensions and stronger tests for my above comment where the base kernel is set directly.

MechCoder commented:
Meh, it was me who suggested that here (https://github.com/scikit-learn/scikit-learn/pull/7070#pullrequestreview-21373896), sorry it's my blunder :/ The way, the kernel code is structured setting theta is similar to `set_params` where allowing modifying constructor params is allowed. Sorry @btabibian!

MechCoder committed:
Fix flake and add my name

MechCoder commented:
@btabibian I initially thought I'll make cosmetic changes before merging but I had to make some non-trivial ones. So I don't think my review is valid.

Let's hope someone else is able to have a look so that we can push this through 0.19

Possible candidates incelude @amueller or @glouppe

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1861,3 +1861,181 @@ def is_stationary(self):
+
+
+class SelectDimensionKernel(Kernel):
+    """Wrapper for applying kernels on selected active dimensions of input data.
```
What does active mean here?

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1861,3 +1861,181 @@ def is_stationary(self):
             self.__class__.__name__, self.gamma, self.metric)
+
+
+class SelectDimensionKernel(Kernel):
```
Is "select dimension" certainly the right name? It doesn't select, and we substitute "select" with "active" and"dimension" with "features" below. Is "kernel" necessary? 

I'm pretty sure the literature just uses statements  like "we use the product of a RBF kernel applied to ...".  Is AppyTo or Applied to sufficient?

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1861,3 +1861,181 @@ def is_stationary(self):
+         `n_active_dims` denoting the indices of the active dimensions.
+
+         For an anisotropic kernel, the length_scale of the kernel and
+         the active_dims should have same length and the indices of the
```
I find this sentence too long and unclear. Same length as what?

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1861,3 +1861,181 @@ def is_stationary(self):
+        self.kernel = kernel
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """Return the kernel k(X, Y) and optionally its gradient on selected
```
Pep257: make this a 1 line summary

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1861,3 +1861,181 @@ def is_stationary(self):
+    kernel : a kernel object.
+        an instantiation of kernel object.
+
+    active_dims: array-like, shape (n_active_dims,) or (n_features,)
```
Is this really the best name for the parameter?

And should we be supporting source object's to extract without copying? Perhaps in next version?

MechCoder replied:
Sorry, I do not understand. What does "source object's to extract without copying" mean?

jnothman replied:
Sorry. Autocorrect bit me again and again on that one.

`slice` objects.

jmetzen replied:
+1 for ``slice`` objects.

jnothman commented on sklearn/gaussian_process/tests/test_kernels.py:
```diff
@@ -319,3 +325,31 @@ def test_repr_kernels():
+    assert_array_equal(sdk1(X), kernel2(X))
+
+    # Check active_dims boolean mask
+    sdk1.set_params(active_dims=[True, True])
```
Is there a similar test with integers?

jnothman commented on sklearn/gaussian_process/tests/test_kernels.py:
```diff
@@ -319,3 +325,31 @@ def test_repr_kernels():
+def test_select_dimension_kernel_properties():
+    kernel1 = RBF(length_scale=[1.0, 2.0])
+    kernel2 = Matern(length_scale=[2.0, 3.0])
+    sdk1 = SelectDimensionKernel(kernel=kernel1, active_dims=[0, 1])
```
I can't see a test where the indices are specified as a numpy array already.

I also can't see a test where the indices are specified out of order.

I also can't see a test where the mask is all false or indices are empty.

jnothman reviewed (changes requested):
A few small things...

MechCoder committed:
Test empty, out of order and np array indices

MechCoder commented:
I've addressed your comments with respect to the tests. The name `active_dims` is borrowed from the GPy package that provides similar functionality, the dimensions to which the kernel is applied are the "active" dimensions and the other are inactive.

I think `AppliedTo` is a better name. Is there consensus?

MechCoder committed:
fix flake 8

jmetzen commented on doc/modules/gaussian_process.rst:
```diff
@@ -608,10 +608,20 @@ References
 
+The :class:`SelectDimensionKernel` can be used to apply a kernel on a set of
+specific features. This can be useful when different kernels, e.g.
+periodic and RBF, are needed to model the different feautres of the dataset.
```
feautres -> features

jmetzen commented on examples/gaussian_process/plot_gpr_dim_selection.py:
```diff
@@ -0,0 +1,78 @@
+# Fixing seed
+rng = np.random.RandomState(0)
+
+# Define a kernel with sum of two RBF kernels applied on individual dimentsion.
```
It is not the sum of two RBF kernels...

jmetzen commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1861,3 +1861,179 @@ def is_stationary(self):
+        K_diag : array, shape (n_samples_X,)
+            Diagonal of kernel k(X, X)
+        """
+        # We have to fall back to slow way of computing diagonal
```
This comment does not make sense here...

jmetzen commented:
The name ``AppliedTo`` doesn't make things clearer for me. I think the term dimension should be in the name. Potentially ``ActiveDimensionsKernel``?

jnothman commented:
that sounds like a name that gp people would be happy with, but I'm not
convinced it's consistent with the rest of our library...

On 3 Jul 2017 12:57 am, "Jan Hendrik Metzen" <notifications@github.com>
wrote:

> The name AppliedTo doesn't make things clearer for me. I think the term
> dimension should be in the name. Potentially ActiveDimensionsKernel?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9179#issuecomment-312496746>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6wG2LJ13CA0PqFJ7Ukxx1BgxFv52ks5sJ6_TgaJpZM4N_vC8>
> .
>

MechCoder commented:
How about `AppliedToDims` or `AppliedToDimensions`?

@jmetzen the term `Select` is generally used for meta-models that just does some sort of selection, for example `SelectKBest` or `SelectFromModel`

raghavrv reviewed (commented):
some minor comments

jnothman commented:
I'm marking this as stalled. @dgilford, would you like to finish up handling the few remaining comments, as it seems you are keen on this feature?

dgilford commented:
> I'm marking this as stalled. @dgilford, would you like to finish up handling the few remaining comments, as it seems you are keen on this feature?

Thanks for updating the status, glad to resurrect this at minimum.

Git, pull requests, and Python/scikit-learn are all relatively new to me, so it would be a bit daunting without some guidance. Are the main comments about documentation? I see conflicts with a few of the files as well, and tbh I'm not sure how to address these.

jnothman commented:
Yes, it might be a somewhat steep learning curve... Most conflicts are easy
to resolve and github (or a desktop git tool such as GitHub Desktop) will
help guide you through each conflict.

You'd need to at a minimum pull this branch into your local git, and push
it to a new branch in a personal fork of scikit-learn, before opening a new
pull request... Already that might be a challenge!

If you'd like to leave it for someone else, you can.

mirko-m commented:
Is it still possible to fix up this pull request? I'm interested in helping out.

nfinan commented:
Will this feature ever make it into the codebase? It seems like it keeps getting stuck trying to merge. This feature is very important for GPR on high-dimensional data!

adrinjalali commented:
We'd need a fresh PR if we're going to merge this in. But we have limited bandwidth for maintaining the GPs.
